File: cli/src/voidrift_cli/commands/verify.py
"""Verify command: Requirements-driven acceptance testing (REQ-VF-3 through REQ-VF-16)."""

from __future__ import annotations

# Tools available to verify-plan agents (consumed by tool_builder.build_local_tools).
AGENT_TOOLS_PLAN: frozenset[str] = frozenset({
    "read_source_file",
    "read_framework_file",
    "write_framework_file",
})

# Tools available to verify-execute agents (consumed by tool_builder.build_local_tools).
AGENT_TOOLS_EXECUTE: frozenset[str] = frozenset({
    "read_framework_file",
    "write_framework_file",
    "read_process_output",
    "http_request",
    "run_command",
    "browser_navigate",
    "browser_screenshot",
    "browser_click",
    "browser_get_text",
})

import json
import re
import signal
import subprocess
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime
from pathlib import Path

from ..agent import build_local_tools
from ..config import get_max_tokens
from ..models import ModelConfig
from ..utils import (
    append_state,
    boot_run,
    check_disk_space,
    check_requirements_exist,
    ensure_voidrift_dir,
    voidrift_dir,
)
from .. import prompts, ui
from ..skills import find_skill
from ..tools.process_manager import start_process, wait_for_ready, stop_all
from ..tools.http_client import clear_sessions
from ..tools.browser import close_all_sessions


# ---------------------------------------------------------------------------
# VERIFY-PLAN.md parser
# ---------------------------------------------------------------------------

_ITEM_RE = re.compile(r"^### (ITEM-\d+)(\s+\[SKIP\])?", re.MULTILINE)



from ._verify_pipeline import (
    _parse_verify_plan,
    _read_arch_field,
    _run_doc_verify,
    _run_plan_agent,
    _run_sub_agent,
    _build_plan_prompt,
    _build_execute_prompt,
    _write_verify_md,
    _update_manifest,
)

def run_verify(worker: ModelConfig) -> int:
    """Execute the verify command (REQ-VF-3 through REQ-VF-16).

    Args:
        worker: Model configuration for both plan and execute agents.

    Returns:
        Exit code (0 for PASS, 1 for FAIL or error).
    """
    check_disk_space()
    d = ensure_voidrift_dir()

    # REQ-VF-P: preflight guard
    if not check_requirements_exist():
        ui.error("REQUIREMENTS.md not found. Run 'voidrift gather <model>' first.")
        return 1

    log, run_id = boot_run("verify")
    ui.header("VoidRift Verify")
    ui.detail(f"Run: {run_id}  Log: {log}")

    process_handle_id: str | None = None
    _interrupted = False

    def _handle_sigterm(signum: int, frame: object) -> None:
        nonlocal _interrupted
        _interrupted = True

    prev_sigterm = signal.signal(signal.SIGTERM, _handle_sigterm)
    prev_sigint = signal.signal(signal.SIGINT, _handle_sigterm)

    try:
        # ── Stage 0: Documentation verification (REQ-VF-17) ─────────────
        ui.stage("Stage 0 — Verifying documentation...")
        from ..tools.filesystem import WriteContext as _WriteContext
        _fs_ctx = _WriteContext(project_dir=d.parent, max_read_lines=worker.max_read_lines)

        doc_bugs_before = set((d / "bugs").glob("DOC-*.md")) if (d / "bugs").exists() else set()
        _doc_verify_ok = _run_doc_verify(worker, d, log, _fs_ctx)
        doc_bugs_after = set((d / "bugs").glob("DOC-*.md")) if (d / "bugs").exists() else set()
        doc_bug_count = len(doc_bugs_after - doc_bugs_before)
        if doc_bug_count:
            ui.warn(f"Documentation: {doc_bug_count} mismatch(es) found. See .voidrift/bugs/DOC-*.md")
        else:
            ui.detail("Documentation consistent with source code.")

        # ── Stage 1: Plan agent ──────────────────────────────────────────
        ui.stage("Stage 1 — Planning test cases...")

        if not _run_plan_agent(worker, d, log, _fs_ctx):
            return 1

        verify_plan_file = d / "VERIFY-PLAN.md"
        if not verify_plan_file.exists():
            ui.error("Plan agent did not write VERIFY-PLAN.md.")
            return 1

        items = _parse_verify_plan(verify_plan_file.read_text())
        if not items:
            ui.error("VERIFY-PLAN.md contains no items.")
            return 1

        testable = [it for it in items if not it["skip"]]
        skipped_items = [it for it in items if it["skip"]]
        ui.detail(f"{len(testable)} test cases, {len(skipped_items)} skipped.")

        # ── Bootstrap ───────────────────────────────────────────────────
        bootstrap_cmd = _read_arch_field(d, "test_bootstrap")
        if bootstrap_cmd and bootstrap_cmd.lower() not in ("none", ""):
            ui.stage("Running test bootstrap...")
            try:
                result = subprocess.run(
                    bootstrap_cmd,
                    shell=True,
                    capture_output=True,
                    text=True,
                    timeout=120,
                )
                if result.returncode != 0:
                    ui.warn(f"Bootstrap exited {result.returncode}: {result.stderr[:200]}")
                else:
                    ui.detail("Bootstrap complete.")
            except subprocess.TimeoutExpired:
                ui.warn("Bootstrap timed out (120s) — continuing.")

        # ── Start product process ────────────────────────────────────────
        startup_cmd = _read_arch_field(d, "startup_command")
        if startup_cmd and startup_cmd.lower() not in ("none", ""):
            ui.stage(f"Starting product: {startup_cmd}")
            handle_result = start_process(startup_cmd)
            try:
                handle_data = json.loads(handle_result)
                process_handle_id = handle_data.get("handle_id")
                ui.detail(f"PID {handle_data.get('pid')} — handle {process_handle_id}")
            except (ValueError, KeyError):
                ui.error(f"Failed to start product: {handle_result}")
                return 1

            # Wait for readiness — default http strategy if startup_command is an http server
            # Fall back to a brief fixed delay if no strategy can be inferred
            ready_result = wait_for_ready(
                process_handle_id,
                strategy="http",
                target="http://localhost:8000/",
                timeout=30,
            )
            if ready_result != "ready":
                # Not fatal — the test cases themselves will detect if the server is down
                ui.warn(f"Readiness check: {ready_result}")

        # ── Stage 2: Concurrent sub-agents ──────────────────────────────
        ui.stage("Stage 2 — Executing test cases...")
        results: list[dict] = []

        # Add skip results immediately
        for item in skipped_items:
            results.append({"item_id": item["item_id"], "status": "skip", "bug_report_path": None})

        if testable:
            (d / "bugs").mkdir(exist_ok=True)
            max_workers = worker.concurrency
            if max_workers == 0:
                max_workers = len(testable)
            max_workers = max(1, max_workers)

            with ThreadPoolExecutor(max_workers=max_workers) as pool:
                futures = {
                    pool.submit(
                        _run_sub_agent,
                        item,
                        run_id,
                        worker,
                        log,
                        process_handle_id,
                    ): item["item_id"]
                    for item in testable
                }
                done_count = 0
                for future in as_completed(futures):
                    item_id = futures[future]
                    try:
                        result = future.result()
                    except Exception as exc:
                        result = {"item_id": item_id, "status": "fail", "bug_report_path": None, "error": str(exc)}
                    results.append(result)
                    done_count += 1
                    status_icon = "✓" if result["status"] == "pass" else "✗"
                    ui.detail(f"  {status_icon} {item_id} ({done_count}/{len(testable)})")

        # ── Update manifest with results (REQ-TM-6, REQ-TM-7) ─────────
        _update_manifest(d, results, run_id)

        # ── Stage 3: Write report ────────────────────────────────────────
        ui.stage("Stage 3 — Writing report...")
        # Sort results by item_id for deterministic output
        results.sort(key=lambda r: r["item_id"])
        verdict = _write_verify_md(d, results, run_id, doc_bug_count=doc_bug_count)

        failed_count = sum(1 for r in results if r["status"] == "fail")
        append_state(
            cmd="verify",
            model_alias=worker.alias,
            summary=(
                f"Verdict: {verdict} — "
                f"{sum(1 for r in results if r['status'] == 'pass')} passed, "
                f"{failed_count} failed, "
                f"{sum(1 for r in results if r['status'] == 'skip')} skipped"
            ),
            files_created=["VERIFY.md"],
        )

        if verdict == "PASS":
            ui.done("Verification passed.")
            return 0
        else:
            ui.error(f"Verification failed — {failed_count} failure(s). See .voidrift/VERIFY.md.")
            return 1

    except KeyboardInterrupt:
        _interrupted = True
        ui.warn("Interrupted.")
        return 1
    finally:
        signal.signal(signal.SIGTERM, prev_sigterm)
        signal.signal(signal.SIGINT, prev_sigint)
        stop_all()
        clear_sessions()
        close_all_sessions()

File: src/commands/verify.ts
/**
 * Verify command: requirements-driven acceptance testing (REQ-VF-3..16).
 */

// OCP contract (REQ-TOOL-8, REQ-ARCH-9)
export const AGENT_TOOLS_PLAN = new Set(["file"]);
export const AGENT_TOOLS_EXECUTE = new Set(["file", "http", "shell", "browser", "process"]);
export const AGENT_TOOLS_DOC = new Set(["file"]);
export const AGENT_TOOL_ACTIONS_PLAN: Record<string, string[]> = { file: ["read", "list"] };
export const AGENT_TOOL_ACTIONS_EXECUTE: Record<string, string[]> = { file: ["read", "write", "list"], http: ["get", "post", "put", "delete"] };
export const AGENT_TOOL_ACTIONS_DOC: Record<string, string[]> = { file: ["read", "write", "list"] };

import { readFileSync, writeFileSync, existsSync, mkdirSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { execSync } from "node:child_process";
import { parse as parseYaml } from "yaml";
import { AgentLoop } from "../agent/loop.js";
import type { ModelInterface } from "../models.js";
import { WriteContext } from "../tools/filesystem.js";
import { buildLocalTools } from "../tools/builder.js";
import { findSkill } from "../skills.js";
import { loadPrompt, buildSystemPrompt } from "../prompts.js";
import { ensureVoidriftDir, bootRun, appendState, checkDiskSpace, checkRequirementsExist, printCommandHeader } from "../utils.js";
import { getMaxTokens } from "../config.js";
import { startProcess, waitForReady, stopAll, readProcessOutput } from "../tools/process.js";
import { clearSessions } from "../tools/http.js";
import { closeAllSessions } from "../tools/browser.js";
import { trackAgentWithStats, elapsedStr } from "./progress.js";
import { resetSharedTracker, displayErrorSummary } from "../errors.js";

export async function runVerify(worker: ModelInterface, onProgress?: (msg: string) => void): Promise<number> {
  const emit = onProgress ?? ((msg: string) => process.stderr.write(msg + "\n"));
  checkDiskSpace();
  resetSharedTracker();
  const d = ensureVoidriftDir();
  if (!checkRequirementsExist()) {
    emit("Error: REQUIREMENTS.md not found. Run 'voidrift gather' first.");
    return 1;
  }

  const [log, runId] = bootRun("verify");
  if (!onProgress) printCommandHeader("verify", worker.config.alias, log);
  const startTime = Date.now();
  const ctx = new WriteContext({ projectDir: join(d, ".."), maxReadLines: worker.config.maxReadLines });
  const frameworkContext = loadPrompt("system", "CONTEXT");

  try {
    // Stage 0: Doc verification (REQ-VF-17)
    const bugsBefore = existsSync(join(d, "bugs")) ? new Set(readdirSync(join(d, "bugs")).filter(f => f.startsWith("DOC-"))) : new Set<string>();
    await _runDocVerify(frameworkContext, worker, d, log, ctx);
    const bugsAfter = existsSync(join(d, "bugs")) ? new Set(readdirSync(join(d, "bugs")).filter(f => f.startsWith("DOC-"))) : new Set<string>();
    const docBugCount = [...bugsAfter].filter(f => !bugsBefore.has(f)).length;

    // Stage 1: Plan agent
    const [planTools, planHandlers] = buildLocalTools("verify-plan", join(d, ".."), ctx);
    const skill = findSkill("QUALITY-QA") ?? "";
    const planPrompt = loadPrompt("verify", "PLAN");
    const planSystem = buildSystemPrompt(frameworkContext, skill, planPrompt);
    emit("▸ Stage 1: Planning test cases...");
    const planAgent = new AgentLoop({
      model: worker, systemPrompt: planSystem, tools: planTools, toolHandlers: planHandlers,
      stream: false, maxTokens: getMaxTokens(worker.config, "verify.plan"), logPath: log, showSpinner: false,
    });
    await trackAgentWithStats(planAgent, loadPrompt("verify", "PLAN-USER"), "  verify plan");

    const verifyPlanPath = join(d, "VERIFY-PLAN.md");
    if (!existsSync(verifyPlanPath)) {
      emit("Error: VERIFY-PLAN.md not produced.");
      return 1;
    }

    const items = parseVerifyPlan(readFileSync(verifyPlanPath, "utf-8"));
    const testable = items.filter(i => !i.skip);
    const skipped = items.filter(i => i.skip);

    // Bootstrap
    const bootstrap = readArchField(d, "test_bootstrap");
    if (bootstrap && bootstrap.toLowerCase() !== "none" && bootstrap !== "") {
      try { execSync(bootstrap, { timeout: 120_000, stdio: "pipe" }); } catch { /* */ }
    }

    // Start product
    const startup = readArchField(d, "startup_command");
    let processHandle: string | undefined;
    if (startup && startup.toLowerCase() !== "none" && startup !== "") {
      const result = startProcess(startup);
      try { processHandle = JSON.parse(result).handle_id; } catch { /* */ }
      if (processHandle) waitForReady(processHandle, "http", "http://localhost:8000/", 30);
    }

    // Stage 2: Concurrent sub-agents
    mkdirSync(join(d, "bugs"), { recursive: true });
    const results: Array<{ itemId: string; status: string; bugPath?: string }> = [];
    for (const item of skipped) results.push({ itemId: item.itemId, status: "skip" });

    emit(`▸ Stage 2: Executing ${testable.length} test cases...`);
    const [execTools, execHandlers] = buildLocalTools("verify-execute", join(d, ".."), ctx);
    const concurrency = worker.config.concurrency ?? 1;
    const taskFns = testable.map(item => async () => {
      const execPrompt = loadPrompt("verify", "EXECUTE");
      const execSystem = buildSystemPrompt(frameworkContext, skill, execPrompt);
      const agent = new AgentLoop({
        model: worker, systemPrompt: execSystem, tools: execTools, toolHandlers: execHandlers,
        stream: false, maxTokens: getMaxTokens(worker.config, "verify.execute"), logPath: log, showSpinner: false,
      });
      try {
        await trackAgentWithStats(agent, item.content, `  ${item.itemId}`);
        const bugPath = join(d, "bugs", `${item.itemId}.md`);
        return { itemId: item.itemId, status: existsSync(bugPath) ? "fail" : "pass", bugPath: existsSync(bugPath) ? bugPath : undefined };
      } catch {
        return { itemId: item.itemId, status: "fail" as const };
      }
    });

    // Run concurrently up to model's concurrency limit (REQ-VF-15)
    const execResults = await _runConcurrent(taskFns, concurrency);
    results.push(...execResults);

    // Stage 3: Report
    results.sort((a, b) => a.itemId.localeCompare(b.itemId));
    const passed = results.filter(r => r.status === "pass").length;
    const failed = results.filter(r => r.status === "fail").length;
    const skipCount = results.filter(r => r.status === "skip").length;
    const verdict = failed === 0 && docBugCount === 0 ? "PASS" : "FAIL";

    const report = [
      `# Verification Report`, ``, `Run: ${runId}`, `Verdict: **${verdict}**`, ``,
      `| Status | Count |`, `|---|---|`,
      `| Passed | ${passed} |`, `| Failed | ${failed} |`, `| Skipped | ${skipCount} |`,
      `| Doc mismatches | ${docBugCount} |`, ``,
      ...results.map(r => `- ${r.itemId}: ${r.status}${r.bugPath ? ` → [bug report](bugs/${r.itemId}.md)` : ""}`),
    ].join("\n");
    writeFileSync(join(d, "VERIFY.md"), report, "utf-8");

    appendState("verify", worker.config.alias,
      `Verdict: ${verdict} — ${passed} passed, ${failed} failed, ${skipCount} skipped`,
      [".voidrift/VERIFY.md"]);

    emit(`\n✓ Verify complete — ${verdict} (${passed} passed, ${failed} failed, ${skipCount} skipped) (${elapsedStr((Date.now() - startTime) / 1000)})`);
    displayErrorSummary(emit);
    return verdict === "PASS" ? 0 : 1;
  } finally {
    stopAll();
    clearSessions();
    closeAllSessions();
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

interface VerifyItem { itemId: string; skip: boolean; content: string }

export function parseVerifyPlan(text: string): VerifyItem[] {
  const items: VerifyItem[] = [];
  const re = /^### (ITEM-\d+)(\s+\[SKIP\])?/gm;
  let match: RegExpExecArray | null;
  const positions: Array<{ itemId: string; skip: boolean; start: number }> = [];
  while ((match = re.exec(text)) !== null) {
    positions.push({ itemId: match[1], skip: !!match[2], start: match.index });
  }
  for (let i = 0; i < positions.length; i++) {
    const end = i + 1 < positions.length ? positions[i + 1].start : text.length;
    items.push({ ...positions[i], content: text.slice(positions[i].start, end).trim() });
  }
  return items;
}

export function readArchField(dir: string, field: string): string | null {
  const archPath = join(dir, "ARCHITECTURE.md");
  if (!existsSync(archPath)) return null;
  const content = readFileSync(archPath, "utf-8");
  if (!content.startsWith("---\n")) return null;
  const end = content.indexOf("\n---\n", 4);
  if (end === -1) return null;
  try {
    const fm = parseYaml(content.slice(4, end));
    return fm?.[field] ? String(fm[field]) : null;
  } catch { return null; }
}

async function _runConcurrent<T>(tasks: Array<() => Promise<T>>, limit: number): Promise<T[]> {
  if (limit <= 1) { const r: T[] = []; for (const t of tasks) r.push(await t()); return r; }
  const results: T[] = new Array(tasks.length);
  let idx = 0;
  const run = async () => { while (idx < tasks.length) { const i = idx++; results[i] = await tasks[i](); } };
  await Promise.all(Array.from({ length: Math.min(limit, tasks.length) }, () => run()));
  return results;
}

async function _runDocVerify(frameworkContext: string, model: ModelInterface, d: string, log: string, ctx: WriteContext): Promise<void> {
  const [tools, handlers] = buildLocalTools("verify-doc", join(d, ".."), ctx);
  const prompt = loadPrompt("verify", "DOC-VERIFY");
  const system = buildSystemPrompt(frameworkContext, undefined, prompt);
  const agent = new AgentLoop({
    model, systemPrompt: system, tools, toolHandlers: handlers,
    stream: false, maxTokens: getMaxTokens(model.config, "verify.execute"), logPath: log, showSpinner: false,
  });
  try { await trackAgentWithStats(agent, loadPrompt("verify", "DOC-VERIFY-USER"), "  doc verification"); } catch { /* */ }
}

// ---------------------------------------------------------------------------
// VerifyCommand class
// ---------------------------------------------------------------------------

import { BaseCommand } from "./base.js";

export class VerifyCommand extends BaseCommand {
  readonly name = "verify";

  async preflight(): Promise<void> {
    await super.preflight();
    if (!checkRequirementsExist()) throw new Error("REQUIREMENTS.md not found. Run 'voidrift gather' first.");
  }

  async execute(): Promise<number> {
    return runVerify(this.model);
  }
}
